### PR TITLE
Added Color, Date and Range parameters

### PR DESCRIPTION
**Pull request by philippjfr**
* Color: Expects an RGB hex string with optional # prefix
* Date: Has to be datetime or np.datetime64 type

The Date parameter could support bounds like Number but I'm not sure how important that is.

### CODE REVIEW

**Comment by philippjfr** [comment]
@jbednar Added bounds validation for Date along with a Range parameter which also has bounds validation.

File: param/__init__.py [hunk 1]
```diff
@@ -1483,3 +1494,117 @@ def get_range(self):
...
 
+class Date(Number):
+    """
+    Date parameter of datetime type.
+    """
+
+    def __init__(self, default=None, allow_None=False, **kwargs):
+        allow_None = True if default is None else allow_None
+        super(Date, self).__init__(default=default, allow_None=allow_None, **kwargs)
+        self._check_value(default)
```
Comment (jbednar) [inline param/__init__.py]: Isn't all of this the default behavior anyway?  I can't see what required there to be a constructor here, but there must be some part of this that doesn't happen already?

File: param/__init__.py [hunk 2]
```diff
@@ -1483,3 +1494,117 @@ def get_range(self):
...
+
+    def _check_value(self,val):
+        if (self.allow_None and val is None):
+            return
+        if not isinstance(val, String.basestring):
+            raise ValueError("Color '%s' only takes a string value."%self._attrib_name)
+        if not re.match('^#?(([0-9a-fA-F]{2}){3}|([0-9a-fA-F]){3})$', val):
+            raise ValueError("Color '%s' only accepts valid RGB hex codes."
+                             % self._attrib_name)
```
Comment (jbednar) [inline param/__init__.py]: Would it be too confusing to accept a numeric triple when it's set, converting it to the RGB hex string required?  The value would change from what the user thought it was; maybe too tricky.

File: param/__init__.py [hunk 3]
```diff
@@ -1483,3 +1494,117 @@ def get_range(self):
...
+
+
+
+class Range(NumericTuple):
+    """Range parameter represents a numeric range with option bounds and softbounds."""
```
Comment (jbednar) [inline param/__init__.py]: "A numeric range with optional bounds and softbounds"

File: param/__init__.py (hunk 2 above)
Comment (philippjfr) [inline param/__init__.py]: > The value would change from what the user thought it was; maybe too tricky.

Right, thought about this but didn't want to magically change the value.

File: param/__init__.py (hunk 1 above)
Comment (philippjfr) [inline param/__init__.py]: Yes, I think this isn't needed, I originally thought it was.

File: param/__init__.py (hunk 2 above)
Comment (jbednar) [inline param/__init__.py]: Ok.  We can always extend it later to support this if we really need it somewhere.

File: param/__init__.py (hunk 1 above)
Comment (philippjfr) [inline param/__init__.py]: It seems a default of None doesn't force allow_None to True on Number types.

Comment (philippjfr) [inline param/__init__.py]: Nevermind, Number defined a default of 0.0.

Comment (jbednar) [inline param/__init__.py]: I just checked, and I think it's all covered, except that in the superclass, _check_value is only run if the value isn't callable.  Avoiding that check is what allows the values to be dynamic, so I *think* deleting the constructor is the right thing to do here.

**Pull request merged**